File: app/src/main/java/com/veganbeauty/app/features/shop/home/ShopHomeFragment.java
package com.veganbeauty.app.features.shop.home;

import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.widget.NestedScrollView;
import androidx.lifecycle.LifecycleOwnerKt;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.viewpager2.widget.ViewPager2;

import com.veganbeauty.app.R;
import com.veganbeauty.app.core.base.RootieFragment;
import com.veganbeauty.app.data.local.entities.ProductEntity;
import com.veganbeauty.app.databinding.ShopHomeBinding;
import com.veganbeauty.app.features.home.HomeHeaderHelper;
import com.veganbeauty.app.features.home.BottomNavHelper;
import com.veganbeauty.app.features.shop.home.models.CategoryUiModel;
import com.veganbeauty.app.features.shop.product.CartBottomSheetFragment;
import com.veganbeauty.app.features.shop.product.CartHelper;
import com.veganbeauty.app.features.shop.product.ChooseQuantityBottomSheet;
import com.veganbeauty.app.features.shop.product.ShopCheckoutFragment;
import com.veganbeauty.app.features.shop.product.ShopVoucherFragment;
import com.veganbeauty.app.features.shop.product.SuggestedProductsBottomSheet;
import com.veganbeauty.app.features.shop.product.detail.ProductDetailLauncher;
import com.veganbeauty.app.features.shop.product.list.ShopListAdapter;
import com.veganbeauty.app.features.shop.product.list.ShopListFragment;

import java.util.ArrayList;
import java.util.List;

public class ShopHomeFragment extends RootieFragment {

    private ShopHomeBinding binding;
    private ShopHomeViewModel viewModel;

    private final Handler sliderHandler = new Handler(Looper.getMainLooper());
    private final Runnable sliderRunnable = new Runnable() {
        @Override
        public void run() {
            if (binding != null) {
                int current = binding.vpBanner.getCurrentItem();
                int total = binding.vpBanner.getAdapter() != null ? binding.vpBanner.getAdapter().getItemCount() : 0;
                if (total > 0) {
                    binding.vpBanner.setCurrentItem((current + 1) % total);
                }
            }
        }
    };

    private final ShopHomeBannerAdapter bannerAdapter = new ShopHomeBannerAdapter();
    private ShopHomeCategoryAdapter categoryAdapter;
    private ShopListAdapter productAdapter;

    private String cartVoucherCode = null;
    private long cartVoucherDiscount = 0L;
    private boolean isHeaderVisible = true;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = ShopHomeBinding.inflate(inflater, container, false);
        com.veganbeauty.app.utils.ProductImageCache.preload(requireContext());
        setupViewModel();
        
        categoryAdapter = new ShopHomeCategoryAdapter(category -> {
            navigateToCategoryList(category);
        });

        productAdapter = new ShopListAdapter(
            product -> {
                navigateToDetail(product);
            },
            product -> {
                if (product.getStock() <= 0) {
                    android.widget.Toast.makeText(requireContext(), "Sản phẩm hiện đã hết hàng", android.widget.Toast.LENGTH_SHORT).show();
                    return;
                }
                ChooseQuantityBottomSheet bottomSheet = new ChooseQuantityBottomSheet(product, new ChooseQuantityBottomSheet.OnQuantitySelectedListener() {
                    @Override
                    public void onAddToCartClick(com.veganbeauty.app.data.local.entities.ProductEntity p, int quantity) {
                        CartHelper.addToCart(requireContext(), LifecycleOwnerKt.getLifecycleScope(getViewLifecycleOwner()), p, quantity);
                    }
                    @Override
                    public void onBuyNowClick(com.veganbeauty.app.data.local.entities.ProductEntity p, int quantity) {
                        com.veganbeauty.app.data.local.entities.CartItemEntity checkoutItem = new com.veganbeauty.app.data.local.entities.CartItemEntity(
                                p.getId(), p.getName(), p.getMainImage(), p.getPrice(), quantity, true
                        );
                        ArrayList<com.veganbeauty.app.data.local.entities.CartItemEntity> list = new ArrayList<>();
                        list.add(checkoutItem);
                        com.veganbeauty.app.features.shop.product.ShopCheckoutFragment checkoutFragment = com.veganbeauty.app.features.shop.product.ShopCheckoutFragment.newInstance(list, "", 0L);
                        getParentFragmentManager().beginTransaction()
                                .replace(R.id.main_container, checkoutFragment)
                                .addToBackStack(null)
                                .commit();
                    }
                });
                bottomSheet.show(getParentFragmentManager(), ChooseQuantityBottomSheet.TAG);
            },
            true
        );

        return binding.getRoot();
    }

    private void setupViewModel() {
        viewModel = new ViewModelProvider(
                requireActivity(),
                ShopHomeViewModelFactory.create(requireContext())
        ).get(ShopHomeViewModel.class);
    }

    @Override
    public void setupUI(View view) {
        binding.vpBanner.setAdapter(bannerAdapter);
        binding.vpBanner.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageSelected(int position) {
                updateBannerDots(position);
                sliderHandler.removeCallbacks(sliderRunnable);
                sliderHandler.postDelayed(sliderRunnable, 4000);
            }
        });

        binding.rvCategories.setLayoutManager(new GridLayoutManager(requireContext(), 2));
        binding.rvCategories.setAdapter(categoryAdapter);

        binding.rvSuggestedProducts.setLayoutManager(new GridLayoutManager(requireContext(), 2));
        binding.rvSuggestedProducts.setAdapter(productAdapter);

        binding.nestedScrollView.setOnScrollChangeListener((NestedScrollView.OnScrollChangeListener) (v, scrollX, scrollY, oldScrollX, oldScrollY) -> {
            View child = v.getChildAt(0);
            if (child != null && scrollY >= child.getMeasuredHeight() - v.getMeasuredHeight() - 400) {
                productAdapter.duplicateItems();
            }
            if (scrollY > 1200) {
                binding.fabBackToTop.setVisibility(View.VISIBLE);
            } else {
                binding.fabBackToTop.setVisibility(View.GONE);
            }

            int dy = scrollY - oldScrollY;
            View header = binding.getRoot().findViewById(R.id.home_header);
            if (header != null) {
                if (scrollY <= 0) {
                    if (!isHeaderVisible) {
                        isHeaderVisible = true;
                        header.animate().translationY(0).setDuration(200).start();
                    }
                } else if (dy > 10 && isHeaderVisible) {
                    isHeaderVisible = false;
                    float translationY = -header.getHeight();
                    if (translationY == 0) {
                        float density = getResources().getDisplayMetrics().density;
                        translationY = -52 * density;
                    }
                    header.animate().translationY(translationY).setDuration(200).start();
                } else if (dy < -10 && !isHeaderVisible) {
                    isHeaderVisible = true;
                    header.animate().translationY(0).setDuration(200).start();
                }
            }
        });

        binding.fabBackToTop.setOnClickListener(v -> binding.nestedScrollView.smoothScrollTo(0, 0));

        binding.btnViewAll.setOnClickListener(v -> {
            SuggestedProductsBottomSheet suggestedBottomSheet = SuggestedProductsBottomSheet.newInstance();
            suggestedBottomSheet.show(getParentFragmentManager(), SuggestedProductsBottomSheet.TAG);
        });

        HomeHeaderHelper.setup(this, binding.getRoot());

        getParentFragmentManager().setFragmentResultListener(ShopVoucherFragment.REQUEST_KEY, getViewLifecycleOwner(), (requestKey, bundle) -> {
            cartVoucherCode = bundle.getString(ShopVoucherFragment.RESULT_VOUCHER_CODE);
            cartVoucherDiscount = bundle.getLong(ShopVoucherFragment.RESULT_VOUCHER_DISCOUNT, 0L);

            CartBottomSheetFragment cartSheet = CartBottomSheetFragment.newInstance(cartVoucherCode, cartVoucherDiscount);
            cartSheet.show(getParentFragmentManager(), CartBottomSheetFragment.TAG);
        });

        BottomNavHelper.setup(this, binding.getRoot(), R.id.nav_shop, tabId -> {
            BottomNavHelper.navigate(this, tabId);
        });
    }

    @Override
    public void observeViewModel() {
        viewModel.banners.observe(getViewLifecycleOwner(), banners -> {
            bannerAdapter.submitList(banners);
            setupBannerDots(banners.size());
        });

        viewModel.categories.observe(getViewLifecycleOwner(), categories -> {
            categoryAdapter.submitList(categories);
        });

        viewModel.suggestedProducts.observe(getViewLifecycleOwner(), products -> {
            if (products == null) return;
            List<ProductEntity> filtered = new ArrayList<>();
            for (ProductEntity p : products) {
                String cat = p.getCategory();
                if (p.isNew() || p.getPrice() >= 500000 || (cat != null && cat.toLowerCase().contains("combo"))) {
                    filtered.add(p);
                }
            }
            productAdapter.submitList(filtered);
        });
    }

    private void navigateToDetail(ProductEntity product) {
        ProductDetailLauncher.open(this, product);
    }

    private void navigateToCategoryList(CategoryUiModel category) {
        ShopListFragment listFragment = new ShopListFragment();
        Bundle args = new Bundle();
        args.putString("CATEGORY_NAME", category.getName());
        listFragment.setArguments(args);

        getParentFragmentManager().beginTransaction()
                .setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out, android.R.anim.fade_in, android.R.anim.fade_out)
                .replace(R.id.main_container, listFragment)
                .addToBackStack(null)
                .commit();
    }

    private void setupBannerDots(int count) {
        if (binding == null) return;
        binding.llBannerDots.removeAllViews();
        for (int i = 0; i < count; i++) {
            View dot = new View(requireContext());
            int size = getResources().getDimensionPixelSize(R.dimen.home_banner_dot_size);
            LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(size, size);
            params.setMarginEnd(getResources().getDimensionPixelSize(R.dimen.home_banner_dot_margin));
            dot.setLayoutParams(params);
            dot.setBackground(createDotDrawable(false));
            binding.llBannerDots.addView(dot);
        }
        updateBannerDots(0);
    }

    private void updateBannerDots(int activeIndex) {
        if (binding == null) return;
        for (int i = 0; i < binding.llBannerDots.getChildCount(); i++) {
            binding.llBannerDots.getChildAt(i).setBackground(createDotDrawable(i == activeIndex));
        }
    }

    private GradientDrawable createDotDrawable(boolean active) {
        GradientDrawable drawable = new GradientDrawable();
        drawable.setShape(GradientDrawable.OVAL);
        drawable.setColor(active ? 0xFFFFFFFF : 0x88FFFFFF);
        return drawable;
    }

    @Override
    public void onResume() {
        super.onResume();
        if (bannerAdapter.getItemCount() > 0) {
            sliderHandler.removeCallbacks(sliderRunnable);
            sliderHandler.postDelayed(sliderRunnable, 4000);
        }
    }

    @Override
    public void onPause() {
        sliderHandler.removeCallbacks(sliderRunnable);
        super.onPause();
    }

    @Override
    public void onDestroyView() {
        sliderHandler.removeCallbacks(sliderRunnable);
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/veganbeauty/app/features/shop/home/ShopHomeFragment.kt
package com.veganbeauty.app.features.shop.home

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.GridLayoutManager
import androidx.room.Room
import com.veganbeauty.app.core.base.RootieFragment
import com.veganbeauty.app.data.local.LocalJsonReader
import com.veganbeauty.app.data.local.RootieDatabase
import com.veganbeauty.app.data.local.entities.ProductEntity
import com.veganbeauty.app.data.repository.ProductRepository
import com.veganbeauty.app.databinding.ShopHomeBinding
import com.veganbeauty.app.features.shop.product.detail.ShopDetailFragment
import com.veganbeauty.app.features.shop.product.list.ShopListAdapter
import com.veganbeauty.app.features.shop.product.list.ShopListFragment
import androidx.viewpager2.widget.ViewPager2
import androidx.viewpager2.widget.CompositePageTransformer
import androidx.viewpager2.widget.MarginPageTransformer
import android.graphics.drawable.GradientDrawable
import android.widget.LinearLayout
import kotlinx.coroutines.launch

class ShopHomeFragment : RootieFragment() {

    private var _binding: ShopHomeBinding? = null
    private val binding get() = _binding!!

    private lateinit var viewModel: ShopHomeViewModel
    
    private val sliderHandler = android.os.Handler(android.os.Looper.getMainLooper())
    private val sliderRunnable = Runnable {
        if (_binding != null) {
            val current = binding.vpBanner.currentItem
            val total = binding.vpBanner.adapter?.itemCount ?: 0
            if (total > 0) {
                binding.vpBanner.currentItem = (current + 1) % total
            }
        }
    }
    
    private val bannerAdapter = ShopHomeBannerAdapter()
    private val categoryAdapter = ShopHomeCategoryAdapter { category ->
        navigateToCategoryList(category)
    }
    private val productAdapter = ShopListAdapter(
        onItemClick = { product ->
            navigateToDetail(product)
        },
        onAddToCartClick = { product ->
            val bottomSheet = com.veganbeauty.app.features.shop.product.ChooseQuantityBottomSheet(
                product = product,
                onAddToCartClick = { p, quantity ->
                    com.veganbeauty.app.features.shop.product.CartHelper.addToCart(requireContext(), lifecycleScope, p, quantity)
                },
                onBuyNowClick = { p, quantity ->
                    val checkoutItem = com.veganbeauty.app.data.local.entities.CartItemEntity(
                        id = p.id,
                        name = p.name,
                        image = p.mainImage,
                        price = p.price,
                        quantity = quantity,
                        isSelected = true
                    )
                    val checkoutFragment = com.veganbeauty.app.features.shop.product.ShopCheckoutFragment.newInstance(arrayListOf(checkoutItem))
                    parentFragmentManager.beginTransaction()
                        .replace(com.veganbeauty.app.R.id.main_container, checkoutFragment)
                        .addToBackStack(null)
                        .commit()
                }
            )
            bottomSheet.show(parentFragmentManager, com.veganbeauty.app.features.shop.product.ChooseQuantityBottomSheet.TAG)
        }
    )

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = ShopHomeBinding.inflate(inflater, container, false)
        setupViewModel()
        return binding.root
    }

    private fun setupViewModel() {
        val db = RootieDatabase.getDatabase(requireContext())
        val repository = ProductRepository(db.productDao(), LocalJsonReader(requireContext()))

        viewModel = ViewModelProvider(this, object : ViewModelProvider.Factory {
            override fun <T : ViewModel> create(modelClass: Class<T>): T {
                @Suppress("UNCHECKED_CAST")
                return ShopHomeViewModel(repository) as T
            }
        })[ShopHomeViewModel::class.java]
    }

    private var cartVoucherCode: String? = null
    private var cartVoucherDiscount = 0L

    override fun setupUI(view: View) {
        // Thiết lập Adapter cho ViewPager Banner
        binding.vpBanner.adapter = bannerAdapter
        
        binding.vpBanner.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                updateBannerDots(position)
                sliderHandler.removeCallbacks(sliderRunnable)
                sliderHandler.postDelayed(sliderRunnable, 4000)
            }
        })

        binding.rvCategories.layoutManager = GridLayoutManager(requireContext(), 2)
        binding.rvCategories.adapter = categoryAdapter

        // Thiết lập Adapter cho Sản phẩm gợi ý (2 cột)
        binding.rvSuggestedProducts.layoutManager = GridLayoutManager(requireContext(), 2)
        binding.rvSuggestedProducts.adapter = productAdapter

        // Set click listener for View All button to show SuggestedProductsBottomSheet
        binding.btnViewAll.setOnClickListener {
            val suggestedBottomSheet = com.veganbeauty.app.features.shop.product.SuggestedProductsBottomSheet.newInstance()
            suggestedBottomSheet.show(parentFragmentManager, com.veganbeauty.app.features.shop.product.SuggestedProductsBottomSheet.TAG)
        }

        // Open Cart when Cart Container is clicked
        binding.flCartContainer.setOnClickListener {
            val cartSheet = com.veganbeauty.app.features.shop.product.CartBottomSheetFragment.newInstance(
                cartVoucherCode,
                cartVoucherDiscount
            )
            cartSheet.show(parentFragmentManager, com.veganbeauty.app.features.shop.product.CartBottomSheetFragment.TAG)
        }

        // Voucher result listener to re-open cart
        parentFragmentManager.setFragmentResultListener(
            com.veganbeauty.app.features.shop.product.ShopVoucherFragment.REQUEST_KEY,
            viewLifecycleOwner
        ) { _, bundle ->
            val code = bundle.getString(com.veganbeauty.app.features.shop.product.ShopVoucherFragment.RESULT_VOUCHER_CODE)
            val discount = bundle.getLong(com.veganbeauty.app.features.shop.product.ShopVoucherFragment.RESULT_VOUCHER_DISCOUNT, 0L)
            cartVoucherCode = code
            cartVoucherDiscount = discount

            val cartSheet = com.veganbeauty.app.features.shop.product.CartBottomSheetFragment.newInstance(
                cartVoucherCode,
                cartVoucherDiscount
            )
            cartSheet.show(parentFragmentManager, com.veganbeauty.app.features.shop.product.CartBottomSheetFragment.TAG)
        }

        // Xử lý click thanh tìm kiếm
        val searchClickListener = View.OnClickListener {
            val searchFragment = com.veganbeauty.app.features.shop.search.ShopSearchFragment()
            parentFragmentManager.beginTransaction()
                .replace(com.veganbeauty.app.R.id.main_container, searchFragment)
                .addToBackStack(null)
                .commit()
        }
        binding.llSearchBar.setOnClickListener(searchClickListener)
        // Make EditText behave like a button
        binding.etSearch.isFocusable = false
        binding.etSearch.isClickable = true
        binding.etSearch.setOnClickListener(searchClickListener)

        // Navigation
        com.veganbeauty.app.features.home.BottomNavHelper.setup(
            fragment = this,
            root = binding.root,
            activeTabId = com.veganbeauty.app.R.id.nav_shop
        ) { tabId -> com.veganbeauty.app.features.home.BottomNavHelper.navigate(this, tabId) }
    }

    override fun observeViewModel() {
        // Quan sát dữ liệu Banner
        lifecycleScope.launch {
            viewModel.banners.collect { banners ->
                bannerAdapter.submitList(banners)
                setupBannerDots(banners.size)
            }
        }

        // Quan sát dữ liệu Category
        lifecycleScope.launch {
            viewModel.categories.collect { categories ->
                categoryAdapter.submitList(categories)
            }
        }

        // Quan sát dữ liệu Sản phẩm gợi ý (Lọc sản phẩm Hot hoặc Mới)
        viewModel.suggestedProducts.observe(viewLifecycleOwner) { products ->
            val filtered = products.filter {
                it.isNew || it.price >= 500000 || it.category.contains("Combo", ignoreCase = true)
            }.take(4)
            productAdapter.submitList(filtered)
        }

        // Observe cart items to update badge count
        lifecycleScope.launch {
            val db = RootieDatabase.getDatabase(requireContext())
            db.cartDao().getAllCartItems().collect { items ->
                val totalQty = items.sumOf { it.quantity }
                if (totalQty > 0) {
                    binding.tvCartBadge.visibility = View.VISIBLE
                    binding.tvCartBadge.text = totalQty.toString()
                } else {
                    binding.tvCartBadge.visibility = View.GONE
                }
            }
        }
    }

    private fun navigateToDetail(product: ProductEntity) {
        val detailFragment = ShopDetailFragment()
        detailFragment.setProduct(product)

        parentFragmentManager.beginTransaction()
            .setCustomAnimations(
                android.R.anim.fade_in,
                android.R.anim.fade_out,
                android.R.anim.fade_in,
                android.R.anim.fade_out
            )
            .replace(com.veganbeauty.app.R.id.main_container, detailFragment)
            .addToBackStack(null)
            .commit()
    }

    private fun navigateToCategoryList(category: com.veganbeauty.app.features.shop.home.models.CategoryUiModel) {
        val listFragment = ShopListFragment()
        listFragment.arguments = Bundle().apply { putString("CATEGORY_NAME", category.name) }
        
        parentFragmentManager.beginTransaction()
            .setCustomAnimations(
                android.R.anim.fade_in,
                android.R.anim.fade_out,
                android.R.anim.fade_in,
                android.R.anim.fade_out
            )
            .replace(com.veganbeauty.app.R.id.main_container, listFragment)
            .addToBackStack(null)
            .commit()
    }

    private fun setupBannerDots(count: Int) {
        if (_binding == null) return
        binding.llBannerDots.removeAllViews()
        repeat(count) {
            val dot = View(requireContext())
            val size = resources.getDimensionPixelSize(com.veganbeauty.app.R.dimen.home_banner_dot_size)
            val params = LinearLayout.LayoutParams(size, size).apply {
                marginEnd = resources.getDimensionPixelSize(com.veganbeauty.app.R.dimen.home_banner_dot_margin)
            }
            dot.layoutParams = params
            dot.background = createDotDrawable(false)
            binding.llBannerDots.addView(dot)
        }
        updateBannerDots(0)
    }

    private fun updateBannerDots(activeIndex: Int) {
        if (_binding == null) return
        for (i in 0 until binding.llBannerDots.childCount) {
            binding.llBannerDots.getChildAt(i).background = createDotDrawable(i == activeIndex)
        }
    }

    private fun createDotDrawable(active: Boolean): GradientDrawable {
        return GradientDrawable().apply {
            shape = GradientDrawable.OVAL
            setColor(if (active) 0xFFFFFFFF.toInt() else 0x88FFFFFF.toInt())
        }
    }

    override fun onResume() {
        super.onResume()
        if (bannerAdapter.itemCount > 0) {
            sliderHandler.removeCallbacks(sliderRunnable)
            sliderHandler.postDelayed(sliderRunnable, 4000)
        }
    }

    override fun onPause() {
        sliderHandler.removeCallbacks(sliderRunnable)
        super.onPause()
    }

    override fun onDestroyView() {
        sliderHandler.removeCallbacks(sliderRunnable)
        super.onDestroyView()
        _binding = null
    }
}
